pkg/image/api/helper.go

```diff
@@ -145,21 +154,108 @@ func ImageWithMetadata(image Image) (*Image, error) {
+		return &TagEvent{
+			Created:              util.Now(),
+			DockerImageReference: ref.String(),
+			Image:                ref.ID,
+		}, nil
+	}
+	return LatestTaggedImage(repo, value)
+}
+
+// DockerImageReferenceForRepository returns a DockerImageReference that represents
+// the ImageRepository or false, if no valid reference exists.
+func DockerImageReferenceForRepository(repo *ImageRepository) (DockerImageReference, error) {
+	spec := repo.Status.DockerImageRepository
+	if len(spec) == 0 {
+		spec = repo.DockerImageRepository
+	}
+	if len(spec) == 0 {
+		return DockerImageReference{}, fmt.Errorf("no possible pull spec for %s/%s", repo.Namespace, repo.Name)
+	}
+	return ParseDockerImageReference(spec)
+}
+
 // LatestTaggedImage returns the most recent TagEvent for the specified image
-// repository and tag.
-func LatestTaggedImage(repo ImageRepository, tag string) (*TagEvent, error) {
-	if _, ok := repo.Tags[tag]; !ok {
-		return nil, fmt.Errorf("image repository %s/%s: tag %q not found", repo.Namespace, repo.Name, tag)
+// repository and tag. Will resolve lookups for the empty tag.
+func LatestTaggedImage(repo *ImageRepository, tag string) (*TagEvent, error) {
+	if len(tag) == 0 {
+		tag = "latest"
+	}
+	// find the most recent tag event with an image reference
+	if repo.Status.Tags != nil {
+		if history, ok := repo.Status.Tags[tag]; ok {
+			for _, item := range history.Items {
+				if len(item.DockerImageReference) > 0 {
+					return &item, nil
+				}
+			}
+		}
 	}
 
-	tagHistory, ok := repo.Status.Tags[tag]
-	if !ok {
-		return nil, fmt.Errorf("image repository %s/%s: tag %q not found in tag history", repo.Namespace, repo.Name, tag)
+	// infer a pull spec given the pull locations - requires the tag
+	// to have a value, for .status.DIR or .DIR to be set, and for
+	// one of those values to be valid.
+	if value, ok := repo.Tags[tag]; ok && len(value) > 0 {
+		ref, err := DockerImageReferenceForRepository(repo)
+		if err != nil {
+			return nil, err
+		}
+		ref.Tag = value
+		return &TagEvent{
+			Created:              util.Now(),
+			DockerImageReference: ref.String(),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("no image recorded for %s/%s:%s", repo.Namespace, repo.Name, tag)
+}
+
+// AddTagEventToImageRepository attempts to update the given image repository with a tag event. It will
+// collapse duplicate entries - returning true if a change was made or false if no change
+// occurred.
+func AddTagEventToImageRepository(repo *ImageRepository, tag string, next TagEvent) bool {
+	if repo.Status.Tags == nil {
+		repo.Status.Tags = make(map[string]TagEventList)
+	}
+
+	tags, ok := repo.Status.Tags[tag]
+	if !ok || len(tags.Items) == 0 {
+		repo.Status.Tags[tag] = TagEventList{Items: []TagEvent{next}}
+		return true
+	}
+
+	previous := &tags.Items[0]
+
+	// image reference has not changed
+	if previous.DockerImageReference == next.DockerImageReference {
+		if next.Image == previous.Image {
+			return false
+		}
+		previous.Image = next.Image
```

Comment:
Can you give an example where the DockerImageReference doesn't change but the Image does?